### DISCUSSION TITLE
tests: Include testing code in code coverage

### DISCUSSION THREAD

**Discussion by webmiche** [xdslproject/xdsl#259] [Polls]
In a recent [PR](https://github.com/xdslproject/xdsl/pull/257) the question came up whether we want to include the testing code in the code coverage.

I am not sure what makes more sense here. From a conceptual perspective, I feel that the code coverage should only contain parts of the library and I don't consider the tests part of that. However, this has the advantage of knowing which tests are actually executed, which can be very insightful as well.

Notice how this influences the code coverage as well. Writing testing code (that obviously gets executed) will always improve coverage, so the exact number is always higher in this case and there is a bias to write a lot of tests (maybe very dumb ones) to increase the code coverage.

**Comment by math-fehr** [comment]
I would argue that if we can do it easily, we should try to cover the testing code as well.
The reason is mostly to have an idea if the tests themselves are sensible, and if they make sense.
Otherwise, what would be the reason in having unreached code in testing?

However, I do not have strong feelings about this, so both options are fine for me. After all, we only "publish" the non-testing code.

**Reply by webmiche** [reply]
The main argument against it is that it inflates the code coverage artificially. Therefore, people could increase coverage by just writing more tests that test the same thing. Obviously, I am not arguing that anyone committing code will do someting malicious, but assuming byzantine nodes in such protocols usually makes sense and covers for mistakes.

**Reply by math-fehr** [reply]
Yeah that makes sense. I'm wondering what peolpe are usually doing?
If no one seems to test this, then that's probably for a good reason then!

**Reply by math-fehr** [reply]
So one argument I have though is that last week I realized I didn't named my function properly, and thus it wasn't tested.
I would like if possible a way to avoid this, though now that we have more people in, these errors should be more often catched in PRs.
Would there be a way to have two coverage metrics, one for the test files, and one for the main code?

**Reply by webmiche** [reply]
I think doing it with tests makes sense early on in a project to check the setup. Later on, when the code base and the test suite expand, not having them in migh tmake more sense.

I think we could generate two metrics. I can look at this...

**Reply by webmiche** [reply]
It's actually very easy: we can use `coverage report --include=./tests/*` with the data already generate on the CI to see the amount of tests run. Notice though that this only includes the pytest tests as coverage only looks at `.py` files, filecheck tests aren't actually executed in the same sense...

But I guess we can see how many filecheck tests have run in the lit report...

**Reply by math-fehr** [reply]
Yeah, if it is easy, then maybe we should do it!
I think mostly to see if the tests are correctly ran or not.

**Reply by webmiche** [reply]
How do you plan to include it into the CI run though? I guess we could try to enforce that it is always >99% or something...

**Reply by math-fehr** [reply]
Well, currently, we just report the coverage on each run, maybe we should just do the same for testing?

**Reply by webmiche** [reply]
In which way report it is my question. There is (a) what is in the logs of the CI and there is (b) the stuff on the coverage page and the PR comment that it updates. Having (a) is pretty much free, I am not sure about (b) though as I haven't set that up personally. Only having (a) though is not very insightful, that's why I suggested to make it such that the MLIR-based CI enforces a testing code coverage of more than 99% (or whatever makes sense) and fails otherwise. The coverage reported in the CI can then only be the actual code coverage without tests.

**Reply by math-fehr** [reply]
Hmm, I'm not sure if having a CI for x% is good or not.
Because essentially, this means we won't look at it until the threshold get passed, and it would be at that moment that we see that some code wasn't executed.
And there is also the fact that operation definitions are not marked as covered sometimes it seems, so we might pass the threshold without making any mistakes.
If we only show the values in the logs, I'm not sure this is good as well, since I doubt we are gonna look at it ;)

**Reply by webmiche** [reply]
Yes, I agree with both points...

It would be cool to be have two measurements in the Codecov comment, one for tests and one for actual code...

**Comment by georgebisbas** [comment]
I am happy either way. I  would vote to keep covering the tests until the project scales...e.g. v2.0...

**Comment by superlopuh** [comment]
The tests are covered!
